add a configuration file

rchakra3 opened the pull request:
Just want to preface this by saying this is the first time I've worked with handling such a large number of configs in go, and this might not be the right approach at all! Please let me know if that's the case! :)

I've used a combination of [viper](https://github.com/spf13/viper) and [validator](https://github.com/go-playground/validator) to achieve the following behavior:
- Read in a toml file (There's an example file with descriptions)
- Set properties from the following sources (in decreasing order of priority):
  - Environment variables
  - properties defined in the config file
  - defaults
- The resulting configuration struct is then validated. 
  - If any properties are provided for a storage backend, then all the required properties should be specified. For example, if a minio endpoint is provided, but secrets are not - a validation error will occur 


I wanted to make sure I was headed down the right path before starting to integrate these changes with the proxy and olympus.

*If* this is the right approach, in it's final state I would expect this to address https://github.com/gomods/athens/issues/190, https://github.com/gomods/athens/issues/284 and to some extent https://github.com/gomods/athens/issues/261

arschles commented on config.full.toml:
```diff
@@ -0,0 +1,122 @@
+#         # Env override: ATHENS_RDBMS_STORAGE_NAME
+#         Name = ""
+#         # Defaults to Global Timeout
+#         TimeoutSeconds = 30
```
@rchakra3 is this an example file (if so, that's awesome by the way!)? if so, can you put a comment indicating that at the top?

rchakra3 commented:
@arschles yes I was hoping to make those changes in smaller increments. 
I did start down that path but realized there would be a TON of small changes all over. 

IMO it might be easier to test/review if we make smaller subsequent PRs for various parts of the codebase and slowly move over to using this instead of env.*. What do you think?

marwan-at-work commented on pkg/config/parse.go:
```diff
@@ -0,0 +1,106 @@
+	return &config, nil
+}
+
+func setDefaultsAndEnvBindings() {
```
This function might be hard to maintain, why don't we use `envconfig` to make our struct more declarative. Something like this: 

```golang
type Config struct {
  EnableCSRFProtection bool
  GoEnv                string         `validate:"required",envconfig:"GO_ENV",default:"development"`
  GoBinary string `validate:"required" envconfig: "GO_BINARY_PATH" default: "go"
}
```
If viper provides the same functionality, then we don't need envconfig. 

And in the initializer, you basically just do 3 steps in a row: 
1. marshal TOML into struct
2. parse envconfig (override TOML if env var exists)
3. validate struct

robjloranger replied:
I agree, it doesn't look like viper provides something like this so we should look into `envconfig`. Viper just uses [pelletier/go-toml](github.com/pelletier/go-toml) to load a map of values.

rchakra3 commented:
Sorry, I've been out for most of the weekend!

envconfig's a great suggestion for maintainability, I'll switch to using that instead of the method.

@robjloranger thanks for the detailed review! :) I'm going to get started on addressing your comments as well.

rchakra3 commented:
I think I've addressed all the initial comments. Going to start adding some tests now!

robjloranger commented on scripts/check_gofmt.sh:
```diff
@@ -5,4 +5,4 @@
 
 GO_FILES=$(find . -iname '*.go' -type f | grep -v /vendor/)   # All the .go files, excluding vendor/
-test -z $(gofmt -s -l $GO_FILES | tee /dev/stderr)
+#test -z $(gofmt -s -l $GO_FILES | tee /dev/stderr)
```
Just forgot to remove commenting from previous commit :slightly_smiling_face:

rchakra3 replied:
That's one last thing I need to figure out :(

gofmt fails on the travis build server for some reason and I haven't been able to figure out why. After spending an hour on it I decided to first validate everything else was working fine. Now, back to figuring that out! :)

rchakra3 replied:
Well. Looks like something I did fixed it. That was odd!

robjloranger replied:
very odd, I wonder what it was

devdinu commented on pkg/config/parse.go:
```diff
@@ -0,0 +1,99 @@
+// ParseConfigFile parses the given file into an athens config struct
+func ParseConfigFile(configFile string) (*Config, error) {
+
+	// attempt to read the given config file
```
`envconfig` supports reading from environment or `yml` file. Any reason to have a `toml` file?

rchakra3 replied:
I didn't think [envconfig](https://github.com/kelseyhightower/envconfig) has support for yaml files. Am I missing something?

No specific reason for toml over yaml, I just personally prefer it for configs.

The idea with this change was to move towards using config files over only environment variables so we could have a few distinct configs (dev/test/prod) that would be easy to switch between and reason about

devdinu replied:
Oops, my bad. yaml support is in viper. If its plain file with `export`, running `go test` in different directory would be simpler. toml works too.

devdinu commented on pkg/config/parse.go:
```diff
@@ -0,0 +1,99 @@
+	return nil
+}
+
+func validateConfig(c Config) error {
```
envconfig has `tag` which does this job `required: true`

devdinu commented on pkg/config/rdbms.go:
```diff
@@ -0,0 +1,7 @@
+// RDBMSConfig specifies the properties required to use an RDBMS as the storage backend
+type RDBMSConfig struct {
+	Name    string `validate:"required" envconfig:"ATHENS_RDBMS_STORAGE_NAME"`
+	Timeout int
```
could rename it to `TimeoutInMs` with the duration unit so we know the usage.

robjloranger replied:
It's in seconds and derived from the global timeout currently.

devdinu replied:
so renaming to`TimeoutSec` would help.

robjloranger replied:
That would be another PR, we do intend to have more than the one timeout. If there is not an issue open yet, can you please open one for that? This PR however should just move us from where we are to using a config file.

devdinu replied:
relevant to this https://github.com/gomods/athens/issues/261

devdinu commented on vendor/github.com/go-playground/locales/currency/currency.go:
```diff
@@ -0,0 +1,308 @@
+package currency
+
+// Type is the currency type associated with the locales currency enum
+type Type int
```
we could remove the vendor directory and add it to `.gitignore` since we've the dependencies with dep. This could go as a separate PR/issue. and others could add more if there's any reason for those checkedin dependencies.

robjloranger replied:
We are currently committing the vendor directory, this prevents upstream VCS changes from breaking builds.

devdinu commented:
Awesome, @rchakra3 infact we could remove all the `env.Vars()` if we're gonna depend on this config package, by exposing these via functions. We could create a issue and track it, if we wanna remove `env`  package.

marwan-at-work commented on pkg/config/storage.go:
```diff
@@ -0,0 +1,74 @@
+
+// StorageConfig provides configs for various storage backends
+type StorageConfig struct {
+	CDN   *CDNConfig   `validate:""`
```
do we need to put `validate:""`?? can we not just omit it? same as below.

rchakra3 replied:
It ends up skipping properties that don't have that tag, and since we want to run the validation for CDNConfig, MongoConfig, etc, we need it here!

marwan-at-work commented on pkg/config/storage.go:
```diff
@@ -0,0 +1,74 @@
+	RDBMS *RDBMSConfig `validate:""`
+}
+
+func setDefaultTimeouts(s *StorageConfig, defaultTimeout int) {
```
do we need this function if we already have `default` struct tags in the Config?

rchakra3 replied:
this sets timeouts for storage backends to the global timeout value if they haven't been specified. Since it's not a static value (it's whatever the global timeout is set to), I couldn't think of a better way to do this!

michalpristas commented on config.example.toml:
```diff
@@ -0,0 +1,144 @@
+    StorageType = "mongo"
+    # Port sets the port the proxy listens on
+    # Env override: PORT
+	Port = ":3000"
```
why theose spaces? is it intentional?

rchakra3 replied:
Wow great catch! Not intentional :)

marwan-at-work commented on pkg/config/cdn.go:
```diff
@@ -0,0 +1,24 @@
+
+// CDNEndpointWithDefault returns CDN endpoint if set
+// if not it should default to clouds default blob storage endpoint e.g
+func (c *CDNConfig) CDNEndpointWithDefault(value *url.URL) *url.URL {
```
Do we need this function? Can we not just work with the URL as a string and not a *ur.URL?

rchakra3 replied:
I've left this in for the `WithDefault` behavior.

Currently the default passed in is the base URI of the storage implementation being used. We could add logic here to default to the Endpoint/URI of the storage config defined in `StorageType`. That would require this config package to be aware of a little more business logic, and if we're doing that we could also just move all the logic in the `GetStorage()` method in `actions/storage.go` to this package as well.

michalpristas commented on pkg/config/proxy.go:
```diff
@@ -0,0 +1,9 @@
+// ProxyConfig specifies the properties required to run the proxy
+type ProxyConfig struct {
+	StorageType           string `validate:"required" envconfig:"ATHENS_STORAGE_TYPE" default:"mongo"`
+	OlympusGlobalEndpoint string `validate:"required" envconfig:"OLYMPUS_GLOBAL_ENDPOINT" default:"olympus.gomods.io"`
```
i think this changed in some recent PR to localhost

rchakra3 replied:
didn't realize, fixed!

marwan-at-work commented on pkg/config/parse.go:
```diff
@@ -0,0 +1,99 @@
+
+var validate *validator.Validate
+
+func getValidator() *validator.Validate {
```
this is a little bit magical. If the *global* variable is not nil then return it, otherwise return a new one. Not sure why we have to set up things this way. We can just initialize `validate` globally on line 12 like this: `var validate = validator.New()` and that's it. Or, we can remove this function as well as the global variable and just initialize a new validator when we wanna use it in other funcs.

rchakra3 replied:
good point :)

marwan-at-work commented on pkg/config/parse.go:
```diff
@@ -0,0 +1,99 @@
+
+func parseConfig(confStr string) (*Config, error) {
+	var config Config
+	if _, err := toml.Decode(confStr, &config); err != nil {
```
you can get rid of all the stuff in lines 41-46 and just use `toml.DecodeFile` and pass the file path :)

rchakra3 replied:
I broke it apart to make it a little easier to test, but I'm happy to use `DecodeFile` too if you think that's better :)

marwan-at-work commented on pkg/config/parse.go:
```diff
@@ -0,0 +1,99 @@
+	// delete invalid storage backend configs
+	// envconfig initializes *all* struct pointers, even if there are no corresponding defaults or env variables
+	// this method prunes all such invalid configurations
+	deleteInvalidStorageConfigs(config.Storage)
```
do we need to delete them? no harm in having them around as opposed to accidentally deleting something we need.

rchakra3 replied:
The problem it solves is that invalid configs (created by envconfig by default) would fail overall validation of the Config struct later. 

I was planning on passing in our logger when I start moving code to use this package so we could log the fact that we're ignoring certain storage configs.

manugupt1 commented on config.example.toml:
```diff
@@ -0,0 +1,144 @@
+# GoBinary returns the path to the go binary to use. This value can be a name of a binary in your PATH, or the full path
+# Defaults to "go"
+# Env override: GO_BINARY_PATH
+GoBinary = "go"
```
Should it default to go1.11rc?

robjloranger replied:
It defaults to `go` because we assume the installation environment has an appropriate version installed.

When not running in a Docker or similar environment where we have an appropriate version of go installed, the operator is expected to set `GO_BINARY_PATH`. Which would currently be `go1.11rc1`, or the full path if the binary is not in the user's `$PATH`.

manugupt1 replied:
Makes sense!

rchakra3 commented:
@marwan-at-work done!

rchakra3 commented:
Ok I just found a bug in this. Please don't merge yet!

rchakra3 commented:
While working on using these configs in the codebase I found an issue - There is no way to know if the value returned by `envconfig` is a default or an Environment variable. 

By extension, when overriding the values provided by the config file with the struct returned by `envconfig` the default value as per envconfig will always override the values provided by the config file. I didn't catch this before because the example config file I was using had the same values as the defaults.

I've added tests to test this behavior with my last commit as well as setting defaults in code. I don't think there's a way to work around this.

marwan-at-work commented on config.example.toml:
```diff
@@ -0,0 +1,154 @@
+        # Defaults to Global Timeout
+        # Env override: MONGO_CONN_TIMEOUT_SEC
+        Timeout = 300
+    [Storage.RDBMS]
```
remove this.

marwan-at-work commented on pkg/config/parse.go:
```diff
@@ -0,0 +1,117 @@
+type Config struct {
+	GoEnv                string         `validate:"required" envconfig:"GO_ENV"`
+	GoBinary             string         `validate:"required" envconfig:"GO_BINARY_PATH"`
+	LogLevel             string         `validate:"required" split_words:"true"`
```
why do we need to split words on log level?

rchakra3 replied:
The env variable is ATHENS_LOG_LEVEL, so that helps with the conversion to a camelcase property

marwan-at-work replied:
@rchakra3 why not just say `evnconfig: "ATHENS_LOG_LEVEL"` so that it's clear exactly what the corresponding ENV VAR is :)

marwan-at-work commented on pkg/config/olympus.go:
```diff
@@ -0,0 +1,20 @@
+	if o == nil {
+		o = &OlympusConfig{}
+	}
+	overrideDefaultStr(&o.StorageType, "memory")
```
i find these super weird, the way we had it with `envy` was actually cleaner. 
I have two suggestions: 
1. Use the `default` struct tags in the struct defintion of configs.
2. Use the TOML file itself to set all the defaults, which makes sense. If you ever wanna change a default, you can set an env var. 

I like 2 better than 1. Do you see a problem in this?

rchakra3 replied:
I would prefer 2 for sure.
It just forces the existence of a config file with all the required values defined, which is what the defaults help with

marwan-at-work replied:
@rchakra3 what do you mean by forcing the existence of a config file? This function only sets a default if I'm not mistaken. 

In any case, the config file is going to be mandatory, the server shouldn't work without it. We can even include it as part of the Go Binary if we _really_ want to.

marwan-at-work commented on pkg/config/parse.go:
```diff
@@ -0,0 +1,117 @@
+	}
+
+	overrideDefaultStr(&config.GoEnv, "development")
+	overrideDefaultStr(&config.GoBinary, "go")
```
just as a random example, the TOML file already sets config.GoBinary to the string "go", so why do we need to do this?

rchakra3 commented:
@marwan-at-work I think I've addressed all your comments!

@arschles I know you had concerns about documentation not lining up - I'll update the docs in #517 once this gets merged because this PR by itself has absolutely no change in behavior.

marwan-at-work commented on pkg/config/storage.go:
```diff
@@ -0,0 +1,66 @@
+	Mongo *MongoConfig `validate:""`
+}
+
+func setStorageTimeouts(s *StorageConfig, defaultTimeout int) {
```
can we not get rid of this function and just use the `default` struct tag instead?

rchakra3 commented on config.example.toml:
```diff
@@ -0,0 +1,147 @@
+    FilterOff = true
+    # Username for basic auth
+    # Env override: BASIC_AUTH_USER
+    BasicAuthUser = "user"
```
done!

robjloranger commented:
Let's get this in if your happy @marwan-at-work, there are some more config changes not yet worked on so we can avoid making Rohan keep this updated.
